core/src/local/persistent_state.rs

```diff
@@ -444,70 +451,77 @@ impl PersistentState {
         let serialized_pk = {
             let pk = KeyType::from(self.indices[0].columns.iter().map(|i| &r[*i]));
             if self.has_unique_index {
-                Self::serialize_prefix(0, &pk)
+                Self::serialize_prefix(&pk)
             } else {
                 // For bases without primary keys we store the actual row values keyed by the index
                 // that was added first. This means that we can't consider the keys unique though, so
                 // we'll append a sequence number.
                 self.indices[0].seq += 1;
-                self.serialize_key(0, &pk, self.indices[0].seq)
+                self.serialize_key(&pk, self.indices[0].seq)
             }
         };
 
         // First insert the actual value for our primary index:
         let pk_value = bincode::serialize(&r).unwrap();
-        batch.put(&serialized_pk, &pk_value).unwrap();
+        let db = self.db.as_ref().unwrap();
+        let value_cf = db.cf_handle("0").unwrap();
+        batch.put_cf(value_cf, &serialized_pk, &pk_value).unwrap();
 
         // Then insert primary key pointers for all the secondary indices:
         for (i, index) in self.indices[1..].iter().enumerate() {
             // Construct a key with the index values, and serialize it with bincode:
             let key = KeyType::from(index.columns.iter().map(|i| &r[*i]));
             // Add 1 to i since we're slicing self.indices by 1..:
-            let serialized_key = self.serialize_key((i + 1) as u32, &key, index.seq);
-            batch.put(&serialized_key, &serialized_pk).unwrap();
+            let index_id = (i + 1).to_string();
+            let cf = db.cf_handle(&index_id).unwrap();
+            let serialized_key = self.serialize_key(&key, index.seq);
+            batch.put_cf(cf, &serialized_key, &serialized_pk).unwrap();
         }
     }
 
     fn remove(&self, batch: &mut WriteBatch, r: &[DataType]) {
         let db = self.db.as_ref().unwrap();
-        let mut do_remove = move |indices: &[PersistentIndex], primary_key: &[u8]| {
-            // First delete the actual primary key row:
-            batch.delete(&primary_key).unwrap();
+        let value_cf = db.cf_handle("0").unwrap();
+        let mut do_remove = move |primary_key: &[u8]| {
+            // Delete the value row first (primary index):
+            batch.delete_cf(value_cf, &primary_key).unwrap();
 
             // Then delete any references that point _exactly_ to
             // that row (i.e. not all matching rows):
-            for (i, index) in indices.iter().enumerate() {
+            for (i, index) in self.indices[1..].iter().enumerate() {
```

Comment:
This might read better as:
```rust
for (i, index) in self.indices.iter().enumerate().skip(1) {
```
So that you can get rid of the `index_id` computation below.